libsolidity/interface/CompilerStack.cpp

```diff
@@ -807,26 +829,37 @@ string CompilerStack::createOnChainMetadata(Contract const& _contract) const
 {
 	Json::Value meta;
 	meta["version"] = 1;
-	meta["language"] = "Solidity";
+	meta["language"] = m_importedSources ? "SolidityAST" : "Solidity";
 	meta["compiler"]["version"] = VersionStringStrict;
 
 	meta["sources"] = Json::objectValue;
 	for (auto const& s: m_sources)
 	{
-		solAssert(s.second.scanner, "Scanner not available");
-		meta["sources"][s.first]["keccak256"] =
-			"0x" + toHex(dev::keccak256(s.second.scanner->source()).asBytes());
-		if (m_metadataLiteralSources)
-			meta["sources"][s.first]["content"] = s.second.scanner->source();
+		if (s.second.scanner)
+		{
+			meta["sources"][s.first]["keccak256"] =
+				"0x" + toHex(dev::keccak256(s.second.scanner->source()).asBytes());
+			if (m_metadataLiteralSources)
+				meta["sources"][s.first]["content"] = s.second.scanner->source();
+			else
+			{
+				meta["sources"][s.first]["urls"] = Json::arrayValue;
+				meta["sources"][s.first]["urls"].append(
+					"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
+				);
+			}
+
+		}
 		else
+			meta[s.first]["keccak256"] = "no source code available";
+		if (m_importedSources)
 		{
-			meta["sources"][s.first]["urls"] = Json::arrayValue;
-			meta["sources"][s.first]["urls"].append(
-				"bzzr://" + toHex(dev::swarmHash(s.second.scanner->source()).asBytes())
-			);
+			solAssert(m_sourceJsons.count(s.first), "no JSON found for source");
```

Comment:
Sorry realised you do support multiple filenames for the AST, so the filename should be present.